### PR TITLE
[W-11172769] implement file system independent matcher.

### PR DESCRIPTION
**Pull request by cristianchan-mulesoft**

### CODE REVIEW

File: src/main/java/org/mule/extension/file/common/api/matcher/PathMatcherPredicate.java [hunk 1]
```diff
@@ -30,13 +29,18 @@ public final class PathMatcherPredicate implements Predicate<String> {
...
    * Creates a new instance using the given pattern
    *
    * @param pattern the pattern to be used to test paths.
+   * @param predicateType if is gonna a match local file system or a remote file system ex:ftp , sftp
+   * @param caseSensitive if the predicateType is a external file system predicate this set the case sensitivity
    */
-  public PathMatcherPredicate(String pattern) {
-    delegate = getPredicateForFilename(pattern);
+
+  public PathMatcherPredicate(String pattern, PredicateType predicateType, final boolean caseSensitive) {
```
Comment (acurci-at-mulesoft) [inline src/main/java/org/mule/extension/file/common/api/matcher/PathMatcherPredicate.java:42]: There is something wrong here. You cannot change the public methods or constructors (mainly when you are working on an API package).
Could you preserve the old signature and overload it with the new parameters?
(Please take into account that the default values should be consistent between classes)

File: src/main/java/org/mule/extension/file/common/api/util/UriUtils.java [hunk 2]
```diff
@@ -118,4 +119,156 @@ public static URI trimLastFragment(URI uri) {
     Integer index = uri.getPath().lastIndexOf(SEPARATOR);
     return index != -1 ? createUri(uri.getPath().substring(0, index)) : null;
   }
+
+  //This implementation is coming from JDK implementation to map a glob patter to a regex
+  public static String toRegexPattern(String globPattern) {
```
Comment (acurci-at-mulesoft) [inline src/main/java/org/mule/extension/file/common/api/util/UriUtils.java:124]: Javadoc

File: src/test/java/org/mule/extension/file/common/api/matcher/PathMatcherPredicateTest.java [hunk 3]
```diff
@@ -0,0 +1,240 @@
...
+
+import static org.junit.Assert.assertFalse;
+import static org.junit.Assert.assertTrue;
+
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.regex.PatternSyntaxException;
+
+import org.junit.Test;
+import org.mule.extension.file.common.api.PredicateType;
```
Comment (acurci-at-mulesoft) [inline src/test/java/org/mule/extension/file/common/api/matcher/PathMatcherPredicateTest.java:18]: The mule imports should be first (put it in line 13))

File: src/test/java/org/mule/extension/file/common/api/matcher/PathMatcherPredicateTest.java [hunk 4]
```diff
@@ -0,0 +1,240 @@
...
+
+public class PathMatcherPredicateTest {
```
Comment (acurci-at-mulesoft) [inline src/test/java/org/mule/extension/file/common/api/matcher/PathMatcherPredicateTest.java:20]: Javadoc

File: src/test/java/org/mule/extension/file/common/api/matcher/PathMatcherPredicateTest.java [hunk 5]
```diff
@@ -0,0 +1,240 @@
...
+
+  public int failures;
+
+  @Test
+  public void globPattern1() {
+    PathMatcherPredicate pmp = new PathMatcherPredicate("*.txt", PredicateType.EXTERNAL_FILE_SYSTEM, true);
```
Comment (acurci-at-mulesoft) [inline src/test/java/org/mule/extension/file/common/api/matcher/PathMatcherPredicateTest.java:26]: I understand that there was no test for the old PathMatcherPredicate constructor, right? Please, add more tests with the old constructor and be sure that is working as before.

**Review by acurci-at-mulesoft — CHANGES_REQUESTED** [review]

File: src/main/java/org/mule/extension/file/common/api/matcher/PathMatcherPredicate.java (hunk 1 above)
Comment (cristianchan-mulesoft) [reply src/main/java/org/mule/extension/file/common/api/matcher/PathMatcherPredicate.java:42]: Yes sure

File: src/main/java/org/mule/extension/file/common/api/matcher/FileMatcher.java [hunk 6]
```diff
@@ -105,6 +106,10 @@
   private Long maxSize;
 
 
+  private PredicateType predicateType = PredicateType.LOCAL_FILE_SYSTEM;
+
+  private boolean caseSensitive = true;
```
[the remark this answers is not in the document]
Comment (cristianchan-mulesoft) [reply src/main/java/org/mule/extension/file/common/api/matcher/FileMatcher.java:111]: That right the parameter is gonna be in the sftp connector and yes too the idea is hiding this property to the other connectors using the file commons

File: src/main/java/org/mule/extension/file/common/api/matcher/FileMatcher.java [hunk 7]
```diff
@@ -105,6 +107,10 @@
   private Long maxSize;
 
 
+  private PredicateType predicateType = LOCAL_FILE_SYSTEM;
+
+  private boolean caseSensitive = true;
```
Comment (acurci-at-mulesoft) [inline src/main/java/org/mule/extension/file/common/api/matcher/FileMatcher.java:112]: I think you should have a public static final with the default value and use it in all the classes that need to set the default value. If you don't do that, you have the risk that changing a value in the future in one class by setting different values in each class.

File: src/main/java/org/mule/extension/file/common/api/matcher/PathMatcherPredicate.java [hunk 8]
```diff
@@ -30,13 +29,29 @@ public final class PathMatcherPredicate implements Predicate<String> {
...
   /**
    * Creates a new instance using the given pattern
    *
    * @param pattern the pattern to be used to test paths.
+   *
    */
+
   public PathMatcherPredicate(String pattern) {
-    delegate = getPredicateForFilename(pattern);
+    delegate = getPredicateForFilename(pattern, PredicateType.LOCAL_FILE_SYSTEM, true);
```
Comment (acurci-at-mulesoft) [inline src/main/java/org/mule/extension/file/common/api/matcher/PathMatcherPredicate.java:42]: Idem. I think the case-sensitive value should come from a constant that has the default value.

**Review by acurci-at-mulesoft — CHANGES_REQUESTED** [review]

**Review by acurci-at-mulesoft — APPROVED** [review]

**Pull request merged**